### PR TITLE
Add Indexer into Forte

### PR DESCRIPTION
**Pull request by AvinashBukkittu**
This PR

- Adds Indexer module into Forte (currently supports EmbeddingBasedIndexer)
  - Users can create an index on a CPU or a GPU by setting "device" in HParams.

- Adds test cases for indexers.

- Introduces two types 
  - Embedding: An abstraction which is used to represent embedding for any entry type. 
    For e.g., this adds support for functionality like sentence.embed() (This PR does not add `embed()` method yet.)
  - Query: An entry type for IR related tasks. This type internally uses `Embedding` to represent queries.

- Tweaks the chatbot example to include the new entries added to the Forte type system.

### CODE REVIEW

File: forte/processors/query_creator.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        if "bot_utterance" in input_pack.pack_names:
+            bot_pack = input_pack.get_pack("bot_utterance")
+            context.append(bot_pack.text)
+
+        for text in context:
+            t = self.tokenizer.encode_text(text)
+            input_ids.append(t[0])
+            segment_ids.append(t[1])
+
+        input_ids = torch.LongTensor(input_ids).to(self.device)
```
Comment (hunterhector) [inline forte/processors/query_creator.py]: These are the steps that's worth abstracting away into an ``Entry.get_embed`` method in the future.

**Comment by hunterhector** [comment]
We now have copyright headers on other files. Could you also add those in the new files?

**Comment by AvinashBukkittu** [comment]
> We now have copyright headers on other files. Could you also add those in the new files?

Fixed in 462e2b1

File: forte/processors/machine_translation_processor.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    def initialize(self, resources: Optional[Resources],
+                   configs: Optional[HParams]):
+        if configs:
+            self.src_language = configs.src_language
+            self.target_language = configs.target_language
+            self.in_pack_name = configs.in_pack_name
+            self.out_pack_name = configs.out_pack_name
+
+    # pylint: disable=no-self-use
+    def _define_input_info(self) -> DataRequest:
```
[the remark this answers is not in the document]
Comment (AvinashBukkittu) [reply forte/processors/machine_translation_processor.py]: Fixed in 1e0aeba

File: forte/processors/query_creator.py (hunk 1 above)
Comment (AvinashBukkittu) [reply forte/processors/query_creator.py]: I guess this relates to https://github.com/asyml/forte/issues/2. This will be fixed as part of that issue.

Comment (hunterhector) [reply forte/processors/query_creator.py]: Yes, that's the plan

**Pull request merged**